pkg/provider/openshiftversionprovider.go

```diff
@@ -0,0 +1,65 @@
+package provider
+
+import (
+	"context"
+
+	"github.com/blang/semver/v4"
+	"k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/kubernetes-telemetry/pkg/internal/openshift"
+	"github.com/kong/kubernetes-telemetry/pkg/types"
+)
+
+const (
+	// OpenShiftVersionKey is a report key used to report the OpenShift version, if any.
+	OpenShiftVersionKey = types.ProviderReportKey("openshift_version")
+	// OpenShiftVersionKind is the OpenShift cluster version kind.
+	OpenShiftVersionKind = Kind(OpenShiftVersionKey)
+)
+
+// NewOpenShiftVersionProvider creates telemetry data provider that will query the
+// configured k8s cluster - using the provided client - to get cluster k8s version.
+func NewOpenShiftVersionProvider(name string, c client.Client, namespace, app string) (Provider, error) {
+	selector := labels.SelectorFromSet(labels.Set(map[string]string{"app": app}))
+	d, err := openshift.NewVersionDetector(c, namespace, selector)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
The provider requires a namespace and an app name, same goes for the workflow. Does it mean that a user will have to explicitly set them? How will this work with KIC where we'd rather expect it to work out of the box rather than forcing our users to pass these parameters (I doubt they will be willing to do so given that's only for our telemetery purposes)?

Maybe it would be possible to implement some heuristic that would guess the namespace containing the OpenShift operator?

Reply:
Downstream code sets them in https://github.com/Kong/kubernetes-ingress-controller/pull/4211. It's developer config surface, not user config surface.

I don't know of a way to "guess" that isn't "know ahead of time" or "list all namespaces and all pods in them and search them for an envvar which looks right". The former is the less than perfect aspect of this check: while it doesn't require new permissions, it does use something that isn't really intended for this purpose.

My thinking was that setting it downstream would make it easier to choose some other target Pod if necessary, but I don't feel too strongly about setting it there versus here. Ultimately the wonkiness doesn't really go away.

Reply:
Oh it seems that's already done 🤦